core/src/splitting/RegressionSplittingRule.cpp

```diff
@@ -94,62 +94,60 @@ void RegressionSplittingRule::find_best_split_value_small_q(const Data& data,
                                                             double& best_decrease,
                                                             const std::vector<double>& responses_by_sample,
                                                             const std::vector<std::vector<size_t>>& samples) {
+  // possible_split_values: the sorted unique split values. Length: num_splits (equal to size_node - 1 if all unique)
```

Comment:
I'm wondering if we actually still need access to `possible_split_values`, or if we could get away with just `sorted_samples`? This would let us skip the `all_values.erase(unique(...` call in `Data::get_all_values`.

Reply:
I discussed with @erikcs offline and he'd prefer to leave this idea for a follow-up PR.

Reply:
Leaving this for now and just recording these notes:

It is possible to to get rid of possible_split_values. Just returning sorted_samples (or the zero to m indices directly) is a bit tricky since in the current setup we cannot fill `counter` and `sums` properly without doing the worst case O(N) fill:  they are allocated with `max_num_unique_values`. If the node contains 1000 samples with the same values, and there are only 100 unique values in the entire data set, it will segfault if we fill it with 1000.

Returning the unique value count too (counting them by doing a final scan, as done in the bucket filling in splitting):

```C++
  num_unique = 1;
  for (size_t i = 0; i < samples.size() - 1; i++) {
    if (get(samples[index[i]], var) < get(samples[index[i + 1]], var)) {
      ++num_unique;
    }
  }
```
is fine but not noticeable faster. Even though it could lead to slightly less constant passes, cache misses from all the extra `.get`'s is costly. That is also the reason why this is slower than copying all the sample values before sorting:

```C++
  std::vector<size_t> sorted_samples(samples);
  std::sort(sorted_samples.begin(), sorted_samples.end(), [&](const size_t& lhs, const size_t& rhs)
    {return get(lhs, var) < get(rhs, var);});
```

Reply:
> in the current setup we cannot fill counter and sums properly without doing the worst case O(N) fill: they are allocated with max_num_unique_values

I'm totally fine if we want to drop this line of investigation, but wanted to note -- could we just overallocate `counter` and `sums` to be of size `size_node`?

Reply:
Yes we could just allocate two new vectors of size_node instead of using those two member arrays.

All of this should be simplified in a follow up, like Skip mentioned in his comment there is a lot of duplication now. Do you have thoughts about how a good design would like like?

Reply:
Sorry, I reviewed too quickly and didn't understand the complexity -- `counter` and `sums` are allocated once per splitting rule. It would be great to file a new issue to start the discussion around how to refactor the splitting rule (I think it will be easier for others to find + follow along that way).